website/src/components/Submission/FileUpload/ColumnMapping.ts

```diff
@@ -0,0 +1,99 @@
+import { type ProcessedFile } from './fileProcessing';
+import type { InputField } from '../../../types/config';
+import stringSimilarity from '../../../utils/stringSimilarity';
+
+export class ColumnMapping {
+    private readonly map: ReadonlyMap<string, string | null>;
+
+    private constructor(map: ReadonlyMap<string, string | null>) {
+        this.map = map;
+    }
+
+    private static getBestMatchingTargetColumn(sourceColumn: string, inputFields: InputField[]): string | null {
+        const [bestMatch, score] = inputFields
+            .map((field): [string, number] => {
+                const score = Math.max(
+                    stringSimilarity(sourceColumn, field.name),
+                    stringSimilarity(sourceColumn, field.displayName ?? ''),
+                );
+                return [field.name, score];
+            })
+            .reduce((maxItem, currentItem) => (currentItem[1] > maxItem[1] ? currentItem : maxItem));
+        return score > 0.8 ? bestMatch : null;
+    }
+
+    /* Create a new mapping with the given columns, doing a best-effort to pre-match columns. */
+    public static fromColumns(sourceColumns: string[], inputFields: InputField[]) {
+        const mapping = new Map();
+        let availableFields = inputFields;
+        sourceColumns.forEach((sourceColumn) => {
+            const bestMatch = this.getBestMatchingTargetColumn(sourceColumn, availableFields);
+            mapping.set(sourceColumn, bestMatch);
+            availableFields = availableFields.filter((field) => field.name !== bestMatch);
+        });
+        return new ColumnMapping(mapping);
+    }
+
+    /* Update the mapping with new source and target columns, trying to keep as much of the 
+       mapping intact as possible. */
+    public update(newSourceColumns: string[], inputFields: InputField[]): ColumnMapping {
```

Comment:
Would it be better/more consistent to call the second parameter `newInputFields`?

Reply:
https://github.com/loculus-project/loculus/pull/3478/commits/d8d1616fa42ab8bc048a70c0ed016472a9ab8b5c